test/OrchardCore.Commerce.Tests.UI.Shortcuts/Controllers/OrderController.cs

```diff
@@ -0,0 +1,112 @@
+﻿using Lombiq.HelpfulLibraries.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc;
+using OrchardCore.Commerce.Abstractions;
+using OrchardCore.Commerce.AddressDataType;
+using OrchardCore.Commerce.Constants;
+using OrchardCore.Commerce.Controllers;
+using OrchardCore.Commerce.Fields;
+using OrchardCore.Commerce.Models;
+using OrchardCore.ContentFields.Fields;
+using OrchardCore.ContentManagement;
+using OrchardCore.ContentManagement.Utilities;
+using OrchardCore.Mvc.Core.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using static OrchardCore.Commerce.Constants.ContentTypes;
+
+namespace OrchardCore.Commerce.Tests.UI.Shortcuts.Controllers;
+
+[DevelopmentAndLocalhostOnly]
+public class OrderController : Controller
+{
+    private readonly Lazy<IPaymentService> _checkoutService;
+    private readonly Lazy<IShoppingCartPersistence> _shoppingCartPersistence;
+    private readonly Lazy<IContentManager> _contentManager;
+    private readonly Lazy<IStripePaymentService> _stripePaymentService;
+
+    public OrderController(
+        Lazy<IPaymentService> checkoutService,
+        Lazy<IShoppingCartPersistence> shoppingCartPersistence,
+        Lazy<IContentManager> contentManager,
+        Lazy<IStripePaymentService> stripePaymentService)
+    {
+        _checkoutService = checkoutService;
+        _shoppingCartPersistence = shoppingCartPersistence;
+        _contentManager = contentManager;
+        _stripePaymentService = stripePaymentService;
+    }
```

Comment:
Don't we usually unpack the Lazy injections here in the constructor?

Reply:
That would defeat their purpose, wouldn't it?

Btw, where do they come from? I just had a discussion with József about `Lazy<T>` injections, and he pointed out that they are not available in OC. See [here](https://github.com/Lombiq/Orchard-Vue.js/pull/81#discussion_r1170245145).

Reply:
I'm not sure these have to be lazy in the first place. I've only ever used Lazy in [`OrchardServices`](https://github.com/Lombiq/Helpful-Libraries/blob/4f8808b3bd7b35d1ebb1914a7b286b437e08fd21/Lombiq.HelpfulLibraries.OrchardCore/DependencyInjection/OrchardServices.cs) so the dependent service only has to resolve the serivces it actually uses ([example in this PR](https://github.com/OrchardCMS/OrchardCore.Commerce/pull/290/files#diff-b8d72cea645b77add9c288d1e2e16fcb7c900df0c010010406ff337c69ffc325L79)) or to avoid cyclic dependency. In both cases Lazy gets unpacked in the ctor.

I think not unpacking Lazy is a micro-optimization and considering that service resolution isn't that expensive there is no need for it. The cognitive simplicity of `_shoppingCartPersistence` actually being an `IShoppingCartPersistence` instead of some wrapper method is more than worth it.

> Btw, where do they come from?

The answer is already in the thread you linked. 😉

Reply:
During development I had cyclic dependency, but it looks like in the finalized logic it is not a problem anymore, so I removed the Lazy injection.